### ISSUE TITLE
close_sinks rule false positive ? 

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#1381]
_From @jaumard on November 20, 2018 12:37_

With the following code: 

```
TransactionReceiptBloc() {
    final loadTransactionController = StreamController<int>(sync: true);
    transactionId = loadTransactionController.sink;
  }

  void close() {
    transactionId.close();
  }
```

dart analyzer still warn about sink not closed, but it is. Or do I miss something ? 

ENV: 
Dart VM version: 2.1.0-dev.9.4.flutter-f9ebf21297 (Thu Nov 8 23:00:07 2018 +0100) on "macos_x64"

_Copied from original issue: dart-lang/sdk#35221_

**Comment by pq** [comment]
_From @lukepighetti on December 9, 2018 21:32_

It looks like your close function is outside of your Bloc class. Can you confirm?

**Comment by pq** [comment]
@jaumard : could you share a more complete example so we can reproduce on our end?

Thanks!

**Comment by jaumard** [comment]
Sure I'll do a small example asap, but the simple way to reproduce can be to simply put the Stream under a List<Stream> and then iterate on the list to close them, the linter will still say it's not closed.
Also for https://github.com/aloisdeniel/built_bloc it will say it's not closed are stream are closed by inherited method and under a List<Stream>

**Comment by rrousselGit** [comment]
I recently faced a similar but different issue with close_sinks:

Consider the following function that returns a `Sink` from somewhere (and correctly internally close it).

```dart
Sink getSink();
```

Then used as such:

```dart
final sink = getSink();
```


This will trigger a `close_sinks` warning on `sink` declaration when there is, in fact, no issue at all.

**Comment by lukepighetti** [comment]
I think `close_sinks` is more confusing than helpful. I usually turn it off. I haven't been able to make sense of when it is helpful and when it is not.

**Comment by felangel** [comment]
I've created a sample app which you can use to reproduce the issue at https://github.com/felangel/bloc-close-sinks-bug.

<img width="802" alt="Screen Shot 2019-10-18 at 11 25 39 AM" src="https://user-images.githubusercontent.com/8855632/67111249-259fd480-f19a-11e9-9588-6c0edd50886c.png">

Hope that helps 👍

**Comment by pq** [comment]
Thanks a million @felangel!

**Comment by felangel** [comment]
@pq is there any update on this?

**Comment by pq** [comment]
Hey @felangel.  Sorry no, I haven't had time to dig in.  Looking at https://github.com/felangel/bloc/pull/572#issuecomment-543679914,  I'm missing some context too.  Are `CounterBlock`s self-closing?  That is, do we simply need to special case them?

**Comment by felangel** [comment]
@pq BlocProvider automatically closes them when it is disposed. You can check out https://github.com/felangel/bloc/blob/926029cae2d7614d38b5a9a8952e36bb59054b02/packages/flutter_bloc/lib/src/bloc_provider.dart#L34 if you want to see the implementation but basically any bloc which is created by a BlocProvider is closed automatically.

**Comment by pq** [comment]
Great.  It sounds like we want to ignore blocks created by `BlocProvider`s.

The initial hurdle is a bit mechanical since I need mock out a bunch of your classes (and now that I'm looking, that's more cumbersome than it should be...)

I'll update this issue w/ progress.

**Comment by rrousselGit** [comment]
As a more general rule, maybe this warning should simply not trigger for sinks that are returned by a function:

```
Sink<T> createSink() {
  return StreamController();
}
```

**Comment by pq** [comment]
Interesting idea @rrousselGit!  I'm definitely for generalizing but need to think this through.

/fyi @bwilkerson (maybe we can chat about this later?)

**Comment by pq** [comment]
Ok, a few follow-ups.  Thanks all for the thoughtful input so far!

@felangel: how do you ensure `BlocProvider` gets disposed?  Is this a place where another rule is in order?

As for @rrousselGit's suggestion, after a conversation w/ @bwilkerson  I'm cooling on it after hearing about experiences with exceptions made to a similar rule implemented in CodePro.  My thinking now is to just update the predicate to be something like:

    isSink(type) && !isBloc(type)

It seems like that would at least hand the `Bloc` case?

If there are other framework classes that are similarly managed we could consider adding a more general purpose marker interface or annotation.

@felangel, @rrousselGit: wdyt?

**Comment by felangel** [comment]
@pq thanks for the quick follow-up! `BlocProvider` is a `StatefulWidget` so it will be disposed by Flutter when it is removed from the widget tree.

Your proposal would address the false positive but it would also remove the warning in places where a `bloc` is manually instantiated (especially outside the context of Flutter).

Ideally, we'd be able to determine where the bloc was instantiated and disable the warning if it was done in the context of BlocProvider's builder.

```dart
BlocProvider(
  builder: (_) => MyBloc(), // sink is automatically closed
  child: MyChild(),
),
```

It would be nice to still have the warning in cases where the bloc is instantiated manually.

```dart
void main() {
  final bloc = MyBloc(); // sink is not automatically closed
}
```

Thoughts?

**Comment by rrousselGit** [comment]
I don't like the exception specific for `flutter_bloc`. Especially considering the solution is not bulletproof, nor specific to `flutter_bloc`.

For example, `BlocProvider` has a secondary `.value` constructor:

And doing this:
```dart
BlocProvider.value(
  value: MyBloc(),
);
```

_should_ trigger the warning because the bloc is never closed.

**Comment by pq** [comment]
Ah, OK.  There are a number of issues here apparently. 🤔 

@rrousselGit that's a great example and points out a short-coming in our current implementation.  I'm pretty sure this won't fire at all currently

```dart
BlocProvider.value(
  value: MyBloc(),
);
```

And nor would this unassigned `IOSink`

```dart
  File('foo.txt').openWrite();
```

(which seems like a bug to me.)




Maybe we should take a step back and try and specify exactly the desired analysis.  I'm wondering if maybe `close_sinks` isn't the right jumping off point for a rule to help steer `Bloc` usage.

Tabling fixing `close_sinks` for the moment, can you describe what we'd want to flag?  What if we had a lint (something like `avoid_unmanaged_blocs`) that identified all instantiations of `Bloc`s outside of a `BlocProvider`'s builder?


```dart
BlocProvider(
  builder: (_) => MyBloc(), // OK
  child: MyChild(),
),

final counterBloc = BlocProvider.of<CounterBloc>(context); // OK

BlocProvider.value(
  value: MyBloc(), // LINT
);

final counterBloc = MyBloc(); // LINT

```

**Comment by felangel** [comment]
@pq I think the snippets you've provided cover the bloc-specific cases and desired behavior 👍

**Comment by rrousselGit** [comment]
If something as specific as `avoid_unmanaged_blocs` is alright, then this discussion should probably include the Flutter team

`Sink` is only an implementation detail in this story.
In the end, it's not about `avoid_unmanaged_blocs` but `widget_build_pure` where we shouldn't create Futures, Bloc, ChangeNotifier, or anything else inside `build`

**Comment by felangel** [comment]
@rrousselGit I don't think this is a Flutter specific topic and it isn't just scoped to a widget's build. In my opinion, keeping a widget's build pure is a separate topic. In this case, maybe `close_blocs` would be a more fitting name if we're tabling a more generic solution.

**Comment by rrousselGit** [comment]
The topic is not Flutter specific, but the proposed solution is.

Take this snippet:

```dart
BlocProvider(
  builder: (_) => MyBloc(), // OK
  child: MyChild(),
),

final counterBloc = BlocProvider.of<CounterBloc>(context); // OK

BlocProvider.value(
  value: MyBloc(), // LINT
);

final counterBloc = MyBloc(); // LINT
```


It falls in the same situation as:

```dart
ChangeNotifierProvider(
  builder: (_) => MyChangeNotifier(), // OK
  child: MyChild(),
),

final counterBloc = Provider.of<CounterBloc>(context); // OK

ChangeNotifierProvider.value(
  value: MyChangeNotifier(), // LINT
);

final counterBloc = MyChangeNotifier(); // LINT
```


Same thing with `FutureProvider`+`Future`,  `StreamProvider`+`Stream`, or custom approaches

**Comment by felangel** [comment]
@rrousselGit can you elaborate on `FutureProvider` + `Future`?

I agree that there is a lot in common between `ChangeNotifier`, `Stream`, and `Bloc` but having more granular rules might be the way to go (`dispose_change_notifiers`, `close_sinks`, `cancel_subscriptions`).

In any case, I also prefer not to have a `bloc`-specific solution because in its current state a `bloc` is a `sink` and as @pq mentioned there are other cases where `close_sinks` is also not working as expected so I would still consider this a bug with the existing `close_sink` implementation as opposed to a feature request for a new rule.

**Comment by felangel** [comment]
@pq are there any updates on this? Thanks!

**Comment by pq** [comment]
Hi @felangel.  Sorry no, this stalled out without a consensus on direction.  Let's re-open the conversation now.  @rrousselGit: any additional thoughts?

**Comment by pq** [comment]
@rrousselGit: any additional thoughts on this one?

**Comment by felangel** [comment]
bump

**Comment by pq** [comment]
Sorry @felangel.  We've been focussed elsewhere.  @rrousselGit: do you  have any additional thoughts on this?

**Comment by rrousselGit** [comment]
Additional, no
But I'm still convinced that the situation shouldn't be flutter_bloc
specific.

IMO the fix should be "if the sink comes from a function, or is returned by
the function that creates it, we don't need to warn"

**Comment by pq** [comment]
Thanks!  See also  parallel conversation in #2113.

**Comment by afucher** [comment]
Hi, any news on that?  🎅

**Comment by pq** [comment]
Sorry @afucher but no...  The big rock right now is NNBD support and migration (#2401).  This one is tricky too and just hasn't gotten prioritized.

**Comment by reynolkb** [comment]
@pq late to the party here, but i'm running into this same error. it sounds like i should just ignore the error for now since there's no way around it? it's only a yellow error and my code is working fine

**Comment by cedvdb** [comment]
This one is triggering for me too:

```dart
class RecordService {
  StreamController<Duration>? _progress$;

  Stream<Duration> record() {
    _progress$ = StreamController();

    Timer.periodic(const Duration(seconds: 1), (timer) {
      if (_progress$ != null) {
        _progress$!.add(Duration(seconds: timer.tick));
      }
        
      else {
        timer.cancel();
      }
    });

    return _progress$!.stream.asBroadcastStream();
  }

  Future<String> stop() async {
    await _progress$!.close();
    _progress$ = null;
    return 'path';
  }
}
```

**Comment by Brads3290** [comment]
For what it's worth, this seems to happen if the `StreamController` is an instance of a generic class too. I've produced it with this example:

No warning:
```dart
class Abc {
  
  // No warning here
  final _streamController = StreamController<String>();

  void dispose() {
    this._streamController.close();
  }
}
```

Warning:
```dart
class Abc<T> {

  // Warning: Close instances of `dart.core.Sink`
  final _streamController = StreamController<String>();

  void dispose() {
    this._streamController.close();
  }
}
```
